azure-iot-device/azure/iot/device/iothub/abstract_clients.py

```diff
@@ -79,6 +86,45 @@ def __init__(self, mqtt_pipeline, http_pipeline):
         self._mqtt_pipeline = mqtt_pipeline
         self._http_pipeline = http_pipeline
 
+        self._inbox_manager = None  # this will be overriden in child class
+        self._receive_type = RECEIVE_TYPE_NONE_SET
+        self._client_lock = threading.Lock()
+
+        # Generic Client handlers
+        self._on_method_request_received = None
+        self._on_twin_desired_properties_patch_received = None
+        self._on_message_received = None
+
+    def _validate_receive_api_invoke(self):
+        """Call this function first in EVERY receive API"""
+        self._client_lock.acquire()
+        if self._receive_type is RECEIVE_TYPE_NONE_SET:
+            # Lock the client to ONLY use receive APIs (no handlers)
+            self._receive_type = RECEIVE_TYPE_API
+        elif self._receive_type is RECEIVE_TYPE_HANDLER:
+            raise exceptions.ClientError(
+                "Cannot use receive APIs - receive handler(s) have already been set"
+            )
+        else:
+            pass
+        self._client_lock.release()
```

Comment:
>self._client_lock.release() [](start = 8, length = 27)

this will not release the lock if an exception is raised.  Recommend using the lock as a context manager.  e..g `with self._client_lock:`

Reply:
Excellent catch

---
In reply to: [456676459](https://github.com/Azure/azure-iot-sdk-python/pull/620#discussion_r456676459) [](ancestors = 456676459)